modules/geo-layers/src/tile-layer/tile-2d-header.js

```diff
@@ -32,13 +36,27 @@ export default class Tile2DHeader {
     return result;
   }
 
+  async _loadData(getTileData) {
+    const {x, y, z, bbox, layerId} = this;
+
+    // Construct unique identifier for request scheduler handle
+    const id = `${layerId}-${x}-${y}-${z}`;
+    const requestToken = await requestScheduler.scheduleRequest(id, () => this.isVisible === true);
+
+    let result;
+    if (requestToken) {
+      result = getTileData({x, y, z, bbox});
+      requestToken.done();
+    }
+    return result;
+  }
+
   loadData(getTileData) {
-    const {x, y, z, bbox} = this;
     if (!getTileData) {
       return;
     }
 
-    this._loader = Promise.resolve(getTileData({x, y, z, bbox}))
+    this._loader = Promise.resolve(this._loadData(getTileData))
```

Comment:
_loadData returns a promise, no need to wrap it again. 

Better you can just make this function async also, and use await.

Reply:
> _loadData returns a promise, no need to wrap it again.

Good point.

> Better you can just make this function async also, and use await.

This is what I was struggling with. `Tileset2D._getTile` calls `Tile2DHeader.loadData`, but doesn't expect a Promise to be returned. So if I make `loadData` `async`, then I need to go back up the chain for higher-level functions to accept promise return values?

Reply:
I also mentioned to @Pessimistress on Slack that I've found the TileLayer annoying to work with because `getTileData` passes _either_ a promise or the resolved data to `renderSubLayers`. So anytime you want to pass multiple values to `renderSubLayers`, you need to do annoying Promise checking like:
https://github.com/visgl/deck.gl/blob/e4af87eacd9e4c790f0a8822bd59611b83f62eb9/modules/geo-layers/src/terrain-layer/terrain-layer.js#L152-L158

Reply:
If you are using `async`, `const result = await data;` should work in both cases 

if you can't fix this at the source.

Reply:
The issue is that `renderSubLayers` can't be `async`.

Reply:
Well in that case the else clause in if statement you quoted wont work either, as those values will evaluate after the function has already returned.

The way the `Tileset3D` class handles this: 
- every time you ask for tiles, it synchronously returns all the tiles already loaded. 
- it also initiates loads of missing tiles for that viewport
- When onTileLoaded callbacks happen, the TIle3DLayer again asks the Tileset3D for all its tiles.

This way renderSubLayers will always get a sync list.

Reply:
Well the tricky part is that `renderSubLayers` can't be async, but it _can_ pass promises as layer props to layers that support it. So for example the SimpleMeshLayer allows `mesh` or `texture` to be `async` props. The issue is that `getTileData` returns a single "opaque" promise, and `renderSubLayers` needs to correctly split it into individual promises for the layer.

Reply:
This problem should be addressed in a separate PR.

In https://github.com/visgl/deck.gl/blob/1798a333023125fe98b97ec7a52676244cf4c4a8/modules/geo-layers/src/tile-layer/tile-layer.js#L158-L162

you can simply change to

```js
const layers = tile.isLoaded && this.renderSubLayers(
```

to avoid sending the promise to `rendrSubLayers`.